### DISCUSSION TITLE
Supported storage providers

### DISCUSSION THREAD

**Discussion by chirayudesai** [seedvault-app/seedvault#180] [Ideas]
This discussion is for the storage providers supported by Seedvault.

Storage providers are where the backup is stored.

Seedvault hands off the encrypted backup data to the storage provider via a set of Android APIs known as Storage Access Framework (SAF).

For more on SAF, see https://developer.android.com/guide/topics/providers/document-provider and https://developer.android.com/training/data-storage/shared/documents-files

When you see other apps in your Files app (such as Nextcloud, Google Drive), or see an app asking for permission to allow them access to a particular folder (or the full internal storage / sdcard / usb storage) - that's SAF.

We currently support Nextcloud, and in theory if an app supports SAF, in theory adding it should be as easy as just editing a xml in the Seedvault source.

Examples of apps that might work:
* Google Drive (See #167)
* Seafile (See #169)

Some apps that won't work:
* Dropbox (does not seem to support SAF)

More ideas welcome. We've already seen some on #127 and #176

**Comment by misanthropos** [comment]
Having installed Lineage 18.1 and tried backup with Nextcloud it seems to be impossible to restore because of missing playstore? It gives me the option to install nextcloud, but since there is no package provider coming with Lineage I have to install one. But after install F-Droid and Nextcloud there is no way to restore anymore.

**Reply by bjo81** [reply]
With installed Play Store, Nextcloud install keeps "pending". it seems somehow it is only installed after finising the setup. So it's also too late for a restore.

**Reply by grote** [reply]
Please see my response here: https://github.com/seedvault-app/seedvault/issues/85#issuecomment-836600630

**Reply by misanthropos** [reply]
Thanks a bundle! 

```adb shell am start-activity -a com.stevesoltys.seedvault.RESTORE_BACKUP```

worked and did restore apps to the phone from nextcloud.
Curious though why it worked without me being asked for the 12 recovery codes. Are they stored permanently on the phone?

**Reply by SimonHeimberg** [reply]
There are more options to restore a backup now. See the FAQ https://github.com/seedvault-app/seedvault/wiki/FAQ#how-to-restore-after-passing-through-setup-wizard , in case of problems visit the (now closed) issue #85 (linked above).

**Comment by jplitza** [comment]
[Android Samba Client](https://f-droid.org/de/packages/com.google.android.sambadocumentsprovider/) supports SAF AFAICT (at least it shows up as a location in the usual files app). While that obviously only works in LAN, that needn't be a problem IMO, and might be much more easily available than Nextcloud (e.g. on every NAS in the world).

**Reply by SimonHeimberg** [reply]
It did support SAF. Unfortunately it not not available in an app store anymore. Somebody started with a replacement, but stopped years ago: https://github.com/marcantony/android-smb-provider

**Comment by akki42** [comment]
[CIFS Documents Provider](https://github.com/wa2c/cifs-documents-provider) allows access to SMB, FTP and FTPS storages via the SAF. Would be great if SeedVault could make use of these...

**Comment by avineas** [comment]
[RSAF](https://github.com/chenxiaolong/RSAF) is a more secure android port of rclone than RoundSync, which was also recently added, so i think it would be good that this amazing app is whitelisted too.

**Comment by SimonHeimberg** [comment]
Where do I see which storage providers are supported?
On https://calyxos.org/docs/guide/apps/seedvault/ I did not get much more info. `a networked service` ... `that supports WebDAV` is quite clear, but `a third party storage provider app like Round Sync` is a bit unspecific. Is this "third party app" _the_ only _one_ or _is there_ any other?
There is no hint in the FAQ https://github.com/seedvault-app/seedvault/wiki/FAQ
The list from source is too technical for me
https://github.com/seedvault-app/seedvault/blob/c863695cd6489d04b0ccbcd8159004029ef83108/app/src/main/res/values/config.xml#L11-L15
and only gets a bit clearer by 
https://github.com/seedvault-app/seedvault/blob/c863695cd6489d04b0ccbcd8159004029ef83108/app/src/main/java/com/stevesoltys/seedvault/ui/storage/StorageOptionFetcher.kt#L31-L35

**Reply by grote** [reply]
Round Sync is an app: https://github.com/newhinton/Round-Sync
Which storage providers work does depend on your ROM though. They are responsible for maintaining that config file above. What is in their is just a suggestion.

**Reply by SimonHeimberg** [reply]
Trying to make my question clearer: Are there other "third party apps" supported? Like [CIFS Documents Provider](#discussioncomment-8500078) (I do not understand if an app needs to be listed in the config file or if having some properties is enough.)

I do understand what works with following supported apps ...
- Round Sync: supports many cloud providers (including Seafile and Google Drive)
- DAVx5: supports webdav servers (including nextcloud)

And I can backup to a local place (if there is enough storage space) and sync this one with an app to where I want it to.
(And backup to a connected usb drive works as well.)

Thank you for the timely answer. I will continue trying :sweat_smile: to ask clearly at start, and maybe I will improve...

**Reply by grote** [reply]
The app needs to be allow-listed in the config file (by the ROM) and also support `ACTION_OPEN_DOCUMENT_TREE` with SAF. 

> And I can backup to a local place (if there is enough storage space) and sync this one with an app to where I want it to.

That is discouraged as we've seen many times users messing this up somehow.
